File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/CareerHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject


class CareerHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.CAREER_CHOOSE,
        RequestTypes.CAREER_STATUS,
        RequestTypes.CAREER_ADD_HEAT,
        RequestTypes.CAREER_ARREST,
        RequestTypes.GET_HUNTER_STATUS
    )

    override val requiresAuth: Boolean = true

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId() ?: return HandlerResult.Err(
            ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION)
        )
        return when (type) {
            RequestTypes.CAREER_CHOOSE -> chooseCareer(ctx, playerId, payload)
            RequestTypes.CAREER_STATUS -> getCareerStatus(ctx, playerId)
            RequestTypes.CAREER_ADD_HEAT -> addHeat(ctx, playerId, payload)
            RequestTypes.CAREER_ARREST -> arrest(ctx, playerId, payload)
            RequestTypes.GET_HUNTER_STATUS -> hunterStatus(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported career request", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun chooseCareer(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val career = payload.str("career")?.lowercase() ?: return HandlerResult.Err(
            ServerError.fromMessage("Career type required (whitehat/blackhat)", ErrorCategory.VALIDATION)
        )
        val result = ctx.services.careerService.chooseCareer(playerId, career)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.CAREER_CHOSEN, mapOf(
                "career" to career,
                "message" to result.message,
                "justiceDelta" to result.justiceDelta,
                "heatDelta" to result.heatDelta
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }

    private suspend fun getCareerStatus(ctx: HandlerContext, playerId: String): HandlerResult {
        val report = ctx.services.careerService.getCareerReport(playerId)
        return HandlerResult.Ok(ResponseTypes.CAREER_STATUS, report)
    }

    private suspend fun addHeat(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val amount = payload["amount"]?.let { (it as? kotlinx.serialization.json.JsonPrimitive)?.content?.toIntOrNull() }
            ?: return HandlerResult.Err(ServerError.fromMessage("Amount required", ErrorCategory.VALIDATION))
        val reason = payload.str("reason") ?: ""
        val result = ctx.services.careerService.addHeat(playerId, amount, reason)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.HEAT_UPDATED, mapOf(
                "heatDelta" to amount,
                "message" to result.message
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }

    private suspend fun hunterStatus(ctx: HandlerContext, playerId: String): HandlerResult {
        val hunters = ctx.services.heatCascadeService.getActiveHunters()
        val isHunted = ctx.services.heatCascadeService.isPlayerHunted(playerId)
        val relevantHunters = hunters.filter { it.targetId == playerId }
        return HandlerResult.Ok(ResponseTypes.HUNTER_STATUS, mapOf(
            "hunters" to relevantHunters.map { mapOf(
                "hunterId" to it.hunterId,
                "targetName" to it.targetName,
                "targetHeat" to it.targetHeat,
                "attemptCount" to it.attemptCount
            )},
            "isHunted" to isHunted
        ))
    }

    private suspend fun arrest(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val targetId = payload.str("targetId")
            ?: return HandlerResult.Err(ServerError.fromMessage("Target player ID required", ErrorCategory.VALIDATION))
        val result = ctx.services.careerService.attemptArrest(playerId, targetId)
        return if (result.success) {
            HandlerResult.Ok(ResponseTypes.ARREST_SUCCESS, mapOf(
                "message" to result.message,
                "justiceDelta" to result.justiceDelta
            ))
        } else {
            HandlerResult.Err(ServerError.fromMessage(result.message, ErrorCategory.VALIDATION))
        }
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/CareerHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class CareerHandler : IHandler
{
    private readonly CareerService _careerService;
    private readonly HeatCascadeService _heatCascadeService;

    public string MessageType => "career_choose";

    public CareerHandler(CareerService careerService, HeatCascadeService heatCascadeService)
    {
        _careerService = careerService;
        _heatCascadeService = heatCascadeService;
    }

    public Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var career = payload?.Str("career");
        if (string.IsNullOrEmpty(career))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Career type required (whitehat/blackhat)" }));

        var result = _careerService.ChooseCareer(playerId, career.ToLowerInvariant());
        if (result.Success)
            return Task.FromResult<IActionResult>(new MessageResult("career_chosen", connectionId, new
            {
                career = career.ToLowerInvariant(),
                message = result.Message,
                justiceDelta = result.JusticeDelta,
                heatDelta = result.HeatDelta
            }));
        return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = result.Message }));
    }

    public Task<IActionResult> HandleStatus(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var report = _careerService.GetCareerReport(playerId);
        return Task.FromResult<IActionResult>(new MessageResult("career_status", connectionId, report));
    }

    public Task<IActionResult> HandleAddHeat(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var amount = payload?.Int("amount");
        if (amount is null || amount.Value <= 0)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Amount required" }));

        var reason = payload?.Str("reason") ?? "";
        var result = _careerService.AddHeat(playerId, amount.Value, reason);
        if (result.Success)
            return Task.FromResult<IActionResult>(new MessageResult("heat_updated", connectionId, new
            {
                heatDelta = amount.Value,
                message = result.Message
            }));
        return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = result.Message }));
    }

    public Task<IActionResult> HandleArrest(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "playerId required" }));

        var targetId = payload?.Str("targetId");
        if (string.IsNullOrEmpty(targetId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Target player ID required" }));

        var result = _careerService.AttemptArrest(playerId, targetId);
        if (result.Success)
            return Task.FromResult<IActionResult>(new MessageResult("arrest_success", connectionId, new
            {
                message = result.Message,
                justiceDelta = result.JusticeDelta
            }));
        return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = result.Message }));
    }

    public async Task<IActionResult> HandleHunterStatus(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var hunters = await _heatCascadeService.GetActiveHunters();
            var isHunted = await _heatCascadeService.IsPlayerHunted(playerId);
            var relevantHunters = hunters.Where(h => h.TargetId == playerId).ToList();

            return new MessageResult("hunter_status", connectionId, new
            {
                hunters = relevantHunters.Select(h => new
                {
                    hunterId = h.HunterId,
                    targetName = h.TargetName,
                    targetHeat = h.TargetHeat,
                    attemptCount = h.AttemptCount
                }),
                isHunted
            });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }
}
